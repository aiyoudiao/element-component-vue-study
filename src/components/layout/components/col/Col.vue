<script setup lang="ts">
import type { ComponentInternalInstance } from 'vue'
import { computed, getCurrentInstance, toRefs, unref, withDefaults } from 'vue'
interface ColType {
  span?: number
  tag?: string
  offset?: number
  pull?: number
  push?: number
  xs?: number | object
  sm?: number | object
  md?: number | object
  lg?: number | object
  xl?: number | object
}

interface StyleType {
  paddingLeft?: string
  paddingRight?: string
}

const props = withDefaults(defineProps<ColType>(), {
  span: 24,
  tag: 'div',
})

const { tag } = toRefs<ColType>(props)
// const slots = useSlots()
const gutter = computed<number>(() => {
  let { parent }: ComponentInternalInstance = getCurrentInstance()
  while (parent && parent?.type?.name !== 'ElRow')
    parent = parent.parent

  return (parent ? parent.props.gutter : 0) as number
})

function useClassesAndStyle(props: ColType) {
  const classList = ['el-col']
  const style: StyleType = {}
  if (unref<number>(gutter)) {
    style.paddingLeft = `${unref<number>(gutter) / 2}px`
    style.paddingRight = style.paddingLeft
  }

  const spanValue = unref(toRefs(props).span) || 0
  if (spanValue)
    classList.push(`el-col-${spanValue}`);

  ['offset', 'pull', 'push'].forEach((prop) => {
    const value = unref(toRefs(props)[prop]) as number || 0
    if (value)
      classList.push(`el-col-${prop}-${value}`)
  });

  ['xs', 'sm', 'md', 'lg', 'xl'].forEach((size) => {
    const result: number | object = unref(toRefs(props)[size])
    const typeName = typeof result
    switch (typeName) {
      case 'number':
        classList.push(`el-col-${size}-${result}`)
        break
      case 'object':
        Object.keys(result).forEach((prop) => {
          if (prop === 'span') {
            classList.push(`el-col-${size}-${result[prop]}`)
            return
          }
          classList.push(`el-col-${size}-${prop}-${result[prop]}`)
        })
    }
  })

  return [classList.join(' '), style]
}

const [classes, style] = useClassesAndStyle(props)
</script>

<script lang="ts">
export default {
  name: 'ElCol',
}
</script>

<template>
  <component :is="unref(tag)" :class="classes" :style="style">
    <slot />
  </component>
</template>

<style lang="scss">
@import '../../../../theme/src/col.scss';
</style>
