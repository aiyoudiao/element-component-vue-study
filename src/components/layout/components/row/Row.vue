<script setup lang="ts">
import { computed, getCurrentInstance, provide, toRefs, unref, withDefaults } from 'vue'

interface RowType {
  tag?: string
  gutter?: number
  type?: string
  justify?: string
  align?: string
}

interface StyleType {
  marginLeft?: string
  marginRight?: string
}

const props = withDefaults(defineProps<RowType>(), {
  tag: 'div',
  gutter: 0,
  justify: 'start',
  align: 'top',
})

const { tag } = toRefs<RowType>(props)

function useClassesAndStyle(props: RowType) {
  const classList = ['el-row']
  const style = computed<StyleType>(() => {
    const result: StyleType = {}
    const gutter = unref<number>(props.gutter)
    if (gutter) {
      result.marginLeft = `-${gutter / 2}px`
      result.marginRight = result.marginLeft
    }
    return result
  })

  const { justify, align, type: typeName } = props

  if (unref(justify) !== 'start')
    classList.push(`is-justify-${unref(justify)}`)

  if (unref(align) !== 'top')
    classList.push(`is-align-${unref(align)}`)

  if (unref(typeName) === 'flex')
    classList.push('el-row--flex')

  provide('el-row', getCurrentInstance())
  return [classList.join(' '), unref(style)]
}

const [classes, style] = useClassesAndStyle(props)
</script>

<script lang="ts">
export default {
  name: 'ElRow',
}
</script>

<template>
  <component :is="unref(tag)" :class="classes" :style="style">
    <slot />
  </component>
</template>

<style lang="scss">
@import '../../../../theme/src/row.scss';
</style>

