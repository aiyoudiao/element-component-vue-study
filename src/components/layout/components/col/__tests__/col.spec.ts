import { mount } from '@vue/test-utils'
import Row from '../../row/Row.vue'
import Col from '../Col.vue'

describe('Col.vue', () => {
  it('Col snapshot', () => {
    const wrapper = mount(Col)
    expect(wrapper.element).toMatchSnapshot()
  })

  describe('Col class', () => {
    it ('contain has class .el-col', () => {
      const wrapper = mount(Col)
      expect(wrapper.find('.el-col').exists()).toBe(true)
    })
    it('contain has class .el-col-{num}', () => {
      const num = 12
      const wrapper = mount(Col, {
        props: {
          span: num,
        },
      })
      expect(wrapper.find(`.el-col-${num}`).exists()).toBe(true)
    })
    it('contain has class .el-col-pull-{num}', () => {
      const num = 3
      const wrapper = mount(Col, {
        props: {
          pull: num,
        },
      })
      expect(wrapper.find(`.el-col-pull-${num}`).exists()).toBe(true)
    })
    it('contain has class .el-col-push-{num}', () => {
      const num = 6
      const wrapper = mount(Col, {
        props: {
          push: num,
        },
      })
      expect(wrapper.find(`.el-col-push-${num}`).exists()).toBe(true)
    })
  })

  it('Col gutter style', () => {
    const count = 10
    const gutter = count * 2
    const wrapper = mount({
      template: `
        <el-row :gutter="${gutter}">
          <el-col>hello world</el-col>
        </el-row>
      `,
      components: {
        'el-row': Row,
        'el-col': Col,
      },
    })
    expect(wrapper.find('.el-col').exists()).toBe(true)
    expect(wrapper.find('.el-col').attributes('style')).toBe(`padding-left: ${count}px; padding-right: ${count}px;`)
  })

  it('The responsive', () => {
    const wrapper = mount({
      template: `
        <el-row :gutter="22">
          <el-col 
            :xs="{span: 22, offset: 2}" 
            :sm="{span: 18, offset: 4}" 
            :md="22" 
            :lg="{span: 20, offset: 2}"
            :xl="18"
          >
            hello world
          </el-col>
        </el-row>
      `,
      components: {
        'el-row': Row,
        'el-col': Col,
      },
    })
    expect(wrapper.find('.el-col').exists()).toBe(true)
    expect(wrapper.find('.el-row').find('.el-col').attributes('style')).toBe('padding-left: 11px; padding-right: 11px;')
    expect(wrapper.find('.el-row').find('.el-col-xs-22').exists()).toBe(true)
    expect(wrapper.find('.el-row').find('.el-col-xs-offset-2').exists()).toBe(true)
    expect(wrapper.find('.el-row').find('.el-col-sm-18').exists()).toBe(true)
    expect(wrapper.find('.el-row').find('.el-col-sm-offset-4').exists()).toBe(true)
    expect(wrapper.find('.el-row').find('.el-col-md-22').exists()).toBe(true)
    expect(wrapper.find('.el-row').find('.el-col-lg-20').exists()).toBe(true)
    expect(wrapper.find('.el-row').find('.el-col-lg-offset-2').exists()).toBe(true)
    expect(wrapper.find('.el-row').find('.el-col-xl-18').exists()).toBe(true)
  })
})
