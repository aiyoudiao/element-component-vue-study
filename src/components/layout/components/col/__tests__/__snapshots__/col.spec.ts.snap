// Vitest Snapshot v1

exports[`Col.vue > Col snapshot 1`] = `
<div
  class="el-col el-col-24"
>
  
  
</div>
`;
